File: androidApp/src/main/java/com/example/justdesserts/androidApp/ui/desserts/form/DessertFormViewModel.kt
package com.example.justdesserts.androidApp.ui.desserts.form

import androidx.lifecycle.ViewModel
import com.apollographql.apollo.api.ApolloExperimental
import com.example.justdesserts.shared.repository.DessertRepository
import com.example.justdesserts.shared.cache.Dessert
import com.example.justdesserts.shared.cache.DessertDetail
import com.example.justdesserts.type.DessertInput

@ApolloExperimental
class DessertFormViewModel constructor(private val repository: DessertRepository): ViewModel() {
    suspend fun getDessert(dessertId: String): DessertDetail? {
        return repository.getDessert(dessertId)
    }
    suspend fun createDessert(dessert: Dessert): Dessert? {
        return repository.newDessert(DessertInput(name = dessert.name, description = dessert.description, imageUrl = dessert.imageUrl))
    }
    suspend fun updateDessert(dessert: Dessert): Dessert? {
        return repository.updateDessert(dessert.id, DessertInput(name = dessert.name, description = dessert.description, imageUrl = dessert.imageUrl))
    }
    suspend fun deleteDessert(dessertId: String): Boolean? {
        return repository.deleteDessert(dessertId)
    }
}

File: iosApp/iosApp/Features/Desserts/Create/DessertFormViewModel.swift
//
//  DessertCreateViewModel.swift
//  iosApp
//
//  Created by Michael Stromer on 1/3/21.
//  Copyright © 2021 orgName. All rights reserved.
//

import Foundation
import SwiftUI
import shared

class DessertFormViewModel: ViewModel, ObservableObject {
    @Published var dessert: Dessert?
    
    let dessertRepository: DessertRepository
    
    init(dessertRepository: DessertRepository) {
        self.dessertRepository = dessertRepository
    }
    
    func createDessert(newDessert: Dessert) {
        dessertRepository.doNewDessert(dessertInput: DessertInput(description: newDessert.description_, imageUrl: newDessert.imageUrl, name: newDessert.name)) { [weak self] (data, error) in
            guard let self = self,
                  let newDessert = data else { return }
            self.dessert = newDessert
        }
    }
    
    func updateDessert(dessert: Dessert) {
        dessertRepository.updateDessert(dessertId: dessert.id, dessertInput: DessertInput(description: dessert.description_, imageUrl: dessert.imageUrl, name: dessert.name)) { [weak self] (data, error) in
            guard let self = self,
                  let updatedDessert = data else { return }
            self.dessert = updatedDessert
        }
    }
    
    func deleteDessert(dessertId: String) {
        dessertRepository.deleteDessert(dessertId: dessertId) { [weak self] (deleted, error) in
            guard let self = self else { return }
            self.dessert = nil
        }
    }
}
